### ISSUE TITLE
Problems with Torch as dependency

### ISSUE THREAD

**Issue by jvanbaarsen** [appsignal/appsignal-elixir#446]
```
== Compilation error in file lib/torch/views/flash_view.ex ==
** (UndefinedFunctionError) function Appsignal.Phoenix.Template.EExEngine.compile/2 is undefined (module Appsignal.Phoenix.Template.EExEngine is not available)
    Appsignal.Phoenix.Template.EExEngine.compile("lib/torch/templates/flash/_flash_messages.html.eex", "_flash_messages.html")
    (phoenix) lib/phoenix/template.ex:349: Phoenix.Template.compile/3
    (phoenix) lib/phoenix/template.ex:160: anonymous fn/4 in Phoenix.Template."MACRO-__before_compile__"/2
    (elixir) lib/enum.ex:1925: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix) expanding macro: Phoenix.Template.__before_compile__/1
    lib/torch/views/flash_view.ex:1: Torch.FlashView (module)
    (elixir) lib/kernel/parallel_compiler.ex:206: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/6
could not compile dependency :torch, "mix compile" failed. You can recompile this dependency with "mix deps.compile torch", update it with "mix deps.update torch" or clean it with "mix deps.clean torch"
```

While this looks like it’s caused by the `Appsignal.phoenix?` check, the issue remains if we remove that check completely.

What’s actually happening is that Torch is trying to compile `https://github.com/infinitered/torch/blob/master/lib/torch/templates/flash/_flash_messages.html.eex`. If your app uses AppSignal’s `Appsignal.Phoenix.Template.EExEngine` set up as a template engine, that will fail, as it hasn’t been compiled yet.

We found that moving the :appsignal dependency above both :plug_cowboy and :torch fixes this issue in our example application.

We’re currently not sure how to fix this issue yet while keeping the current setup working for older installs. @danielberkompas do you have some pointers that could help us fix this issue?

Originally reported here: https://app.intercom.io/a/apps/yzor8gyw/inbox/inbox/2788926/conversations/20441137880 (Private intercom link)

**Comment by danielberkompas** [comment]
Hmm, I don't have any pointers at the moment. I will look into this and get back to you.

**Comment by danielberkompas** [comment]
@jvanbaarsen Do you know what version of Torch this happens with? Does it still happen with `2.0.0-rc.1`?

**Comment by epilgrim** [comment]
yes, it still happens on   
```
"torch": {:hex, :torch, "2.0.0-rc.1", "edd2e3b3cb1434d592cbfbde33c5be5692b33f6e38912c9ece6fa7545ef83097", [:mix], [{:filtrex, "~> 0.4.1", [hex: :filtrex, repo: "hexpm", optional: false]}, {:phoenix, "~> 1.3", [hex: :phoenix, repo: "hexpm", optional: false]}, {:phoenix_html, "~> 2.10", [hex: :phoenix_html, repo: "hexpm", optional: false]}, {:scrivener_ecto, ">= 1.2.1", [hex: :scrivener_ecto, repo: "hexpm", optional: false]}], "hexpm"},
```

**Comment by jvanbaarsen** [comment]
@danielberkompas I tested this with the latest version of Torch, and I believe one version older Elixir (Due to no support for Ecto 3.0 in Torch)

**Comment by NobbZ** [comment]
This bug is mainly due to a race in dependency compilation.

If and only if `phoenix` gets compiled before `appsignal` the `EExEngine` module will be compiled into code because of how the conditional compilation resolves:

https://github.com/appsignal/appsignal-elixir/blob/develop/lib/appsignal/phoenix/template/eex_engine.ex#L1

This should be fixable by unconditionally compiling the module. It won't hurt the runtime if its there and not used.

**Comment by gjaldon** [comment]
@NobbZ that module expects `Phoenix.Template.Engine` so I think it may fail if Phoenix wasn't compiled before appsignal.

**Comment by NobbZ** [comment]
Does it call anything from that module during compile time?

**Comment by schnittchen** [comment]
This is probably the root cause of https://github.com/appsignal/appsignal-elixir/issues/319 as well.

For several years now, every now and then I get this error. From these two issues I understand that the root cause is not obvious at first (see this issue's topic).

There are probably three ways to go for this:
* follow @NobbZ analysis https://github.com/appsignal/appsignal-elixir/issues/446#issuecomment-524746014, and either unconditionally compile the engine, or figure out under what exact circumstances Phoenix gets compiled before Appsignal (this might break in the future with changes to mix)
* break out all phoenix related stuff into its own hex package, `appsignal-phoenix`

**Comment by schnittchen** [comment]
Just ran into this again.

**Comment by jeffkreeftmeijer** [comment]
Hey @schnittchen,

This is most likely caused by the same issue as #319, indeed. We're keeping this one open to make sure we've solved the problems for Torch when we fix that one. We've looked into unconditionally compiling the engines, but haven't found a way to get that to work without throwing compile time warnings.

**Comment by jeffkreeftmeijer** [comment]
This should be solved in #319 with the release of 2.0. Please reopen if you’re still running into issues with AppSignal and Torch. Thanks, all!
